### PR TITLE
Test merge

### PR DESCRIPTION
**Pull request by arvidn**

### CODE REVIEW

**Comment by d-komarov** [comment]
OK, I'l take a look tomorrow.

File: src/torrent.cpp [hunk 1]
```diff
@@ -979,17 +980,15 @@ bool is_downloading_state(int const st)
...
+		if (m_share_mode)
 		{
-			m_file_priority.clear();
-			m_file_priority.resize(m_torrent_file->num_files(), dont_download);
+			std::size_t const num_files = valid_metadata()
+				? std::size_t(m_torrent_file->num_files())
+				: m_file_priority.size();
+			// in share mode, all pieces have their priorities initialized to
+			// dont_download
+			prioritize_files(aux::vector<download_priority_t, file_index_t>(num_files, dont_download));
```
Comment (d-komarov) [inline src/torrent.cpp]: I guess this line might be exceeding length limit.

File: src/torrent.cpp [hunk 2]
```diff
@@ -5017,15 +5016,46 @@ bool is_downloading_state(int const st)
...
+				// initialize pad files to priority 0
+				if (files[i] > dont_download && fs && fs->pad_file_at(i))
+					files[i] = dont_download;
+				else if (files[i] > top_priority)
+					files[i] = top_priority;
+			}
+
+			if (fs)
+			{
+				files.resize(fs->num_files(), default_priority);
```
Comment (d-komarov) [inline src/torrent.cpp]: This line should be placed before loop, in order to reset pad files priorities.

File: src/torrent.cpp [hunk 3]
```diff
@@ -5078,33 +5096,33 @@ bool is_downloading_state(int const st)
 		// setting file priority on a torrent that doesn't have metadata yet is
 		// similar to having passed in file priorities through add_torrent_params.
 		// we store the priorities in m_file_priority until we get the metadata
-		file_storage const& fs = m_torrent_file->files();
-		if (index < file_index_t(0) || (valid_metadata() && index >= fs.end_file()))
+		if (index < file_index_t(0)
```
Comment (d-komarov) [inline src/torrent.cpp]: Is `file_index_t` type allowed to have negative values? What does it represent with negative value?

File: src/torrent.cpp [hunk 4]
```diff
@@ -5078,33 +5096,33 @@ bool is_downloading_state(int const st)
...
 		if (m_storage)
 		{
 			ADD_OUTSTANDING_ASYNC("file_priority");
 			m_ses.disk_thread().async_set_file_priority(m_storage
-				, m_file_priority, std::bind(&torrent::on_file_priority, shared_from_this(), _1, _2));
+				, std::move(new_priority), std::bind(&torrent::on_file_priority, shared_from_this(), _1, _2));
+		}
+		else
+		{
+			m_file_priority = new_priority;
```
Comment (d-komarov) [inline src/torrent.cpp]: Using `swap` is more effective then copy assignment. Using move assignment is also fine.

File: src/torrent.cpp (hunk 3 above)
Comment (arvidn) [reply src/torrent.cpp]: it depends on what you mean by "allowed". It refers to an invalid index, similar to an index thats >= the number of files

Comment (d-komarov) [reply src/torrent.cpp]: Ah, got it.

File: test/test_torrent.cpp [hunk 5]
```diff
@@ -81,48 +103,32 @@ void test_running_torrent(std::shared_ptr<torrent_info> info, std::int64_t file_
 		return;
 	}
 
-	aux::vector<download_priority_t, file_index_t> ones(std::size_t(info->num_files()), 1_pri);
```
[the remark this answers is not in the document]
Comment (d-komarov) [reply test/test_torrent.cpp]: No problem.

**Comment by arvidn** [comment]
@d-komarov I updated this patch too

**Comment by d-komarov** [comment]
Yes. Though some comments still need your attention.

**Comment by d-komarov** [comment]
Looks good to me now.

**Pull request merged**